### PR TITLE
Refactor and share asset state filtering logic across catalog + graph

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

In a follow up PR I will also make them share the actual filter creation logic too but considered out of scope for this PR since it's already hefty. The main goal was to add code location filtering to the catalog but in the process I ended up refactoring the filtering logic so I wouldn't have to duplicate it in 3 places.

## How I Tested These Changes

Shadow-dagit elementl,
tested global graph, group graph, and "all assets" catalog

Screenshot 2024-04-01 at 1 36 23 PM

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/asset-graph/AssetGraphExplorer.tsx [hunk 1]
```diff
@@ -65,100 +65,43 @@ import {EmptyDAGNotice, EntirelyFilteredDAGNotice, LoadingNotice} from '../pipel
...
-      filters: Filters;
-      setFilters: React.Dispatch<React.SetStateAction<Filters>>;
-    }
-  | {filters?: null; setFilters?: null};
-
 type Props = {
   options: GraphExplorerOptions;
   setOptions?: (options: GraphExplorerOptions) => void;
 
+  assetFilterState?: AssetFilterState;
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/asset-graph/AssetGraphExplorer.tsx:75]: I know it's a stupid nit but I didn't realize at first that this was more than the `type FIlters` above, and actually contains both the filters and their setters.  I wonder if just calling it `assetFilters` (leaving off "state") would make it more clear that it's a full interface?

Also I really like how this cleans up this page, it's so much more readable with all the state stuff moved out.

File: js_modules/dagster-ui/packages/ui-core/src/assets/AssetsCatalogTable.tsx [hunk 2]
```diff
@@ -219,33 +168,14 @@ export const AssetsCatalogTable = ({
...
   const allAssetGroupOptions = useAssetGroupSelectorsForAssets(pathMatches);
   const allComputeKindTags = useAssetKindTagsForAssets(pathMatches);
   const allAssetOwners = useAssetOwnersForAssets(pathMatches);
 
   const groupsFilter = useAssetGroupFilter({
-    assetGroups: allAssetGroupOptions,
-    visibleAssetGroups: filters.groups,
-    setGroupFilters,
+    allAssetGroups: allAssetGroupOptions,
+    assetGroups: filters.groups,
```
Comment (bengotow) [inline js_modules/dagster-ui/packages/ui-core/src/assets/AssetsCatalogTable.tsx:177]: This is a nice rename!

**Review by bengotow — APPROVED** [review]
This is awesome and so much cleaner than it was before -- I actually hadn't realized how much the filtering had grown until you pulled it out and it cut like 30 lines off some of these files 😂 

This is also super timely, the new asset health overview page calls for all these filters on yet another surface 🤦  I think the new page can use useAssetDefinitionFilterState since it most closely resembles the catalog table view, but will be nice to merge the hooks too!

File: js_modules/dagster-ui/packages/ui-core/src/asset-graph/AssetGraphExplorer.tsx (hunk 1 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/asset-graph/AssetGraphExplorer.tsx:75]: The thing is I started refactoring the actual hooks too and realized that I would actually want to call that abstraction AssetFilters, while this one is just the state and updating the state.

**Pull request merged**